packages/xo-server/src/xapi/mixins/patching.js

```diff
@@ -205,121 +318,16 @@ export default {
-    // - sorted requirements (If A requires B and C, then C cannot require B)
-    // For each requested patch:
-    // - throw if not found
-    // - throw if already installed
-    // - ignore if already in installable (may have been added because of requirements)
-    // - if paid patch on free host: either ignore (listing all the patches) or throw (patch is requested)
-    // - throw if conflicting patches installed
-    // - throw if conflicting patches in installable
-    // - throw if one of the requirements is not found
-    // - push its required patches in installable
-    // - push it in installable
-    const installable = []
-    forEach(requestedPatches, id => {
-      const patch = all[id]
-      if (patch === undefined) {
-        throw new Error(`patch not found: ${id}`)
-      }
-
-      if (installed[id] !== undefined) {
-        if (getAll) {
-          return
-        }
-        throw new Error(`patch already installed: ${patch.name} (${id})`)
-      }
-
-      if (find(installable, { id }) !== undefined) {
-        return
-      }
-
-      if (patch.paid && freeHost) {
-        if (getAll) {
-          return
-        }
-        throw new Error(
-          `requested patch ${patch.name} (${id}) requires a XenServer license`
-        )
-      }
-
-      let conflictId
-      if (
-        (conflictId = find(
-          patch.conflicts,
-          conflictId => installed[conflictId] !== undefined
-        )) !== undefined
-      ) {
-        if (getAll) {
-          log.debug(
-            `patch ${patch.name} (${id}) conflicts with installed patch ${conflictId}`
-          )
-          return
-        }
-        throw new Error(
-          `patch ${patch.name} (${id}) conflicts with installed patch ${conflictId}`
-        )
-      }
-
-      if (
-        (conflictId = find(patch.conflicts, conflictId =>
-          find(installable, { id: conflictId })
-        )) !== undefined
-      ) {
-        if (getAll) {
-          log.debug(`patches ${id} and ${conflictId} conflict with eachother`)
-          return
-        }
-        throw new Error(
-          `patches ${id} and ${conflictId} conflict with eachother`
-        )
-      }
-
-      // add requirements
-      forEach(patch.requirements, id => {
-        const requiredPatch = all[id]
-        if (requiredPatch === undefined) {
-          throw new Error(`required patch ${id} not found`)
-        }
-        if (!installed[id] && find(installable, { id }) === undefined) {
-          if (requiredPatch.paid && freeHost) {
-            throw new Error(
-              `required patch ${requiredPatch.name} (${id}) requires a XenServer license`
-            )
-          }
-          installable.push(requiredPatch)
-        }
-      })
-
-      // add itself
-      installable.push(patch)
-    })
-
-    return installable
-  },
+  _listInstallablePatches,
+  _listInstallablePatchesDebounced,
 
   // high level
   listMissingPatches(hostId) {
```

Comment:
Why not debounce this one directly?

Reply:
In case we want different debounce delays.

Reply:
Oh but `delay` can be a function so we could use that if we need to...